### PR TITLE
Chat bubble scaling (+ shouting and whispering)

### LINKED ISSUE #2560: Chat messages not containing letters are interpreted as yelling
## Description

When a player types a string not containing letters (for example "...") the chat window interprets it as yelling.

### Detailed Bug Report
This is caused by the Chat.Process.cs only checking that the player's message converted to uppercase is identical. (Lines 68-90)
```
var toUpperCheck = message.ToUpper(CultureInfo.InvariantCulture);

if (message.Contains("!") && toUpperCheck != message){
	verb = "exclaims,";
}

```

### Steps to Reproduce
Please enter the steps to reproduce the bug or behaviour:  

1.  Type ... into chat.
2.  Unity_2020-01-07_23-10-44

### LINKED ISSUE #2567: Theme handler overrides opacities of themed elements
# Description
ThemeHandler.cs currently overrides the alpha value of themed elements.

Line 50:
```
        public void SetTheme(ThemeConfig config)
        {
            switch (targetElement)
            {
                case UIElement.Image:
                    image.color = config.imageColor;
                    break;
                case UIElement.Text:
                    text.color = config.textColor;
                    break;
            }
        }
```

This currently happens with all themes, all the time.

Chatbubbles, for example, have opacity in their prefab. This gets overwritten in the theme.

# Solutions

1. Stop ThemeHandler.cs from touching the alpha value
2. Specify alpha values in themes (probably the better but harder solution)

# Notes

Currently working on fixing this in my current PR.

### PR DESCRIPTION
**Pull request by daalta**
# Purpose
- This PR makes it so chat bubbles always stay the same size regardless of the current zoom. They are no longer unreadable when too far zoomed out / in. 
- Chat bubbles are increased or decreased in size if players whisper or shout (see notes).
- Chat bubble max / min size is clamped (https://github.com/unitystation/unitystation/pull/2561/commits/1e08d61607b831c49087db23d9b4de68e01cce86)
- Fixes #2560
- Fixes #2567
- Also adresses #2269, but only for text bubbles. The font sizes for speaking, whispering and shouting were manually checked to be readable, though there's still aliasing.

_Update: Chat bubbles look slightly different from the video with a smaller border, tail and bug-fixed opacity  (https://github.com/unitystation/unitystation/pull/2561/commits/1e08d61607b831c49087db23d9b4de68e01cce86):_
vlc_2020-01-09_01-46-51

### Notes:
- Bubbles stay the same size by scaling themselves according to the camera zoom.
- Chat bubbles are not visible in darkness. This was previously not an issue because chat bubbles were much smaller and were rarely far away from the player. However they can now reach into occluded parts of the station when zoomed out very far.
- Players can shout by typing something in capital letters or ending their sentence with two exclamation marks ("!!")
- Players can whisper by starting their sentence with a number sign ("#"). The number sign at the start of the message is removed.
- The detection of the player whispering / shouting is **not** consistent with Chat.Process.cs and does not use ChatModifier. The syntax for shouting also currently differs from the chat window (awaiting feedback from reviewers). PlayerChatBubble.cs could be rewritten to make use of them in the future.
- Added placeholder comments for clowns having their own chat bubble.
- Feedback appreciated.

### Please make sure you have followed the self checks below before submitting a PR:

- [x] Code is sufficiently commented (But the rest PlayerChatBubble.cs wasn't / isn't well commented)
- [x] Code is indented with tabs and not spaces
- [x] The PR does not include any unnecessary .meta, .prefab or <b>.unity (scene) changes</b>
- [x] The PR does not bring up any new compile errors
- [x] The PR has been tested in editor
- [x] The PR has been tested in multiplayer (Locally on one editor and one regular build)

### What changes to chat bubbles I'd like to see in the future:
- Rich text with TextMeshPro for more interesting chat messages (Comic Sans font for clowns)
- ~~ Smaller "bubble" around the text ~~
- Tweaks to bubble size, text size, max message length, etc. (Adressed in https://github.com/unitystation/unitystation/pull/2561/commits/1e08d61607b831c49087db23d9b4de68e01cce86)
- Fix to chat bubbles being obscured by shadows (or is that a feature..?)

### CODE REVIEW

File: UnityProject/Assets/Scripts/Chat/Chat.Process.cs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ﻿using UnityEngine;
 using System.Text.RegularExpressions;
 using System.Globalization;
+using System.Linq;
```
Comment (daalta) [inline UnityProject/Assets/Scripts/Chat/Chat.Process.cs:4]: Hope adding Linq is OK, haven't seen it discussed much. Felt like it was (in this case) easier to code & understand than regex .

File: UnityProject/Assets/Scripts/Chat/Chat.Process.cs [hunk 2]
```diff
@@ -83,13 +84,14 @@ public partial class Chat
 		}
 
 		var verb = "says:";
-		var toUpperCheck = message.ToUpper(CultureInfo.InvariantCulture);
+		// Yelling if all letters are capitalized and message contains at least one letter.
+		bool toUpperCheck = message == message.ToUpper(CultureInfo.InvariantCulture) && message.Any(System.Char.IsLetter);
```
Comment (daalta) [inline UnityProject/Assets/Scripts/Chat/Chat.Process.cs:88]: Messages like "..." are no longer interpreted as yelling.

**Comment by daalta** [comment]
The latest commit added several additional features and bug fixes, please check the commit message for more information. https://github.com/unitystation/unitystation/pull/2561/commits/1e08d61607b831c49087db23d9b4de68e01cce86

**Comment by chairbender** [comment]
Nice, thank you for recording a gif, the community always appreciates those and they can be annoying to do. Reviewing...

File: UnityProject/Assets/Scripts/UI/PlayerChatBubble/PlayerChatBubble.cs [hunk 3]
```diff
@@ -10,6 +12,54 @@
 /// </summary>
 public class PlayerChatBubble : MonoBehaviour
 {
+	[SerializeField]
```
Comment (chairbender) [inline UnityProject/Assets/Scripts/UI/PlayerChatBubble/PlayerChatBubble.cs:15]: great to have so many editor fields, and they're all appropriately private 👍

File: UnityProject/Assets/Scripts/UI/PlayerChatBubble/PlayerChatBubble.cs [hunk 4]
```diff
@@ -40,6 +101,12 @@ void OnDisable()
 
 	void Update()
 	{
+		// Update scale of the chat bubble.
+		if (showingDialogue)
+		{
+			UpdateChatBubbleSize();
+		}
+		// TODO Add comment to the following code (or remove it). Bubbles seem to work fine without it.
 		if (transform.eulerAngles != Vector3.zero)
```
Comment (chairbender) [inline UnityProject/Assets/Scripts/UI/PlayerChatBubble/PlayerChatBubble.cs:110]: this forces them upright when a matrix is rotating, it's needed I am pretty sure

File: UnityProject/Assets/Scripts/Chat/Chat.Process.cs (hunk 1 above)
Comment (chairbender) [reply UnityProject/Assets/Scripts/Chat/Chat.Process.cs:4]: linq is totally fine as long as it doesn't cause performance issues. I prefer to use it in a lot of cases for readability as well.

**Review by chairbender — APPROVED** [review]
thanks for making chat bubbles even nicer! There's only 2 minor issues but this is working + an improvement so I'll go ahead and merge this. I created #2569 to track these issues. It's up to you if you want to work on it. It's pretty minor so no shame in ignoring it for now.

**Comment by daalta** [comment]
Thank you for the review. You mentioned two minor issues, but I am not seeing the second one.

**Pull request merged**